File: mtproto/src/main/java/com/walhalla/mtprotolist/manager/MUtils.java
//package com.walhalla.mtprotolist.manager;
//
//import android.os.Handler;
//import android.widget.Toast;
//
//import com.walhalla.mtprotolist.Config;
//import com.walhalla.mtprotolist.entity.MtprotoProxy;
//import com.walhalla.mtprotolist.utils.QueryFormat;
//import com.walhalla.mtprotolist.webproxy.ProxyInfo;
//import com.walhalla.ui.DLog;
//
//import okhttp3.MediaType;
//import okhttp3.OkHttpClient;
//import okhttp3.Request;
//import okhttp3.RequestBody;
//import okhttp3.Response;
//
//import org.json.JSONArray;
//import org.json.JSONException;
//import org.json.JSONObject;
//
//import java.util.List;
//import java.util.concurrent.Executor;
//import java.util.concurrent.Executors;
//
//public class MUtils {
//    private final Handler h;
//    private final Executor executor = Executors.newCachedThreadPool(); // Пул потоков
//
//    public void getProxyType(ProxyInfo info, Callback<ProxyInfo> callback) {
//        executor.execute(() -> {
//            OkHttpClient client = new OkHttpClient();
//            Request request = new Request.Builder()
//                    .url(info.proxyUrl)
//                    .build();
//            try {
//                Response response = client.newCall(request).execute();
//                String responseBody = response.body().string();
//
//                if (responseBody.contains("Glype")) {
//                    h.post(() -> {
//                        info.type = "Glype";
//                        callback.onResponse(info);
//                    });
//                }
//                else if (responseBody.contains("PHP-Proxy")) {
//                    h.post(() -> {
//                        info.type = "PHP-Proxy";
//                        callback.onResponse(info);
//                    });
//                }
//                else if (responseBody.contains("PHProxy")) {
//                    h.post(() -> {
//                        info.type = "PHProxy";
//                        callback.onResponse(info);
//                    });
//                }
//                else {
//                    h.post(()->{
//                        info.type = "Other";
//                        callback.onResponse(info);
//                    });
//                }
//            } catch (Exception e) {
//                h.post(()->{
//                    callback.onError(e.getClass().getSimpleName());
//                });
//            }
//        });
//    }
//
//
//    public interface Callback<T> {
//        void onError(String error);
//
//        void onResponse(T serverList);
//    }
//
//    public MUtils(Handler handler) {
//        this.h = handler;
//    }
//
//    private final OkHttpClient client = new OkHttpClient();
//    private final String urlCheckIpBatch = "http://ip-api.com/batch";
//
//    public void getIpInfo(final List<MtprotoProxy> serverList, Callback<List<MtprotoProxy>> callback) {
//        executor.execute(() -> {
//            try {
//                JSONArray jsonArray = QueryFormat.getArray0(serverList);
//                MediaType JSON = MediaType.parse("application/json; charset=utf-8");
//                RequestBody requestBody = RequestBody.create(JSON, jsonArray.toString());
//
//                Request request = new Request.Builder()
//                        .url(urlCheckIpBatch)
//                        .post(requestBody)
//                        .build();
//
//                Response response = client.newCall(request).execute();
//
//                if (response.isSuccessful()) {
//                    // Обработка успешного ответа
//                    final JSONArray jsonResponse = new JSONArray(response.body().string());
//                    h.post(() -> {
//                        if (setIpInfo(jsonResponse, serverList)) {
//                            callback.onResponse(serverList);
//                        }
//                    });
//                } else {
//                    // Обработка ошибки
//                    final String errorMessage = response.message();
//                    h.post(() -> callback.onError(errorMessage));
//                }
//            } catch (Exception e) {
//                e.printStackTrace();
//                final String errorMessage = e.getMessage();// Обработка исключений
//                h.post(new Runnable() {
//                    @Override
//                    public void run() {
//                        callback.onError(errorMessage);
//                    }
//                });
//            }
//        });
//    }
//
//    public void getIpInfoProxyInfo(
//            final List<ProxyInfo> serverList,
//            Callback<List<ProxyInfo>> callback) {
//        executor.execute(() -> {
//            try {
//                JSONArray jsonArray = QueryFormat.getQueryArray(serverList);
//                MediaType JSON = MediaType.parse("application/json; charset=utf-8");
//                RequestBody requestBody = RequestBody.create(JSON, jsonArray.toString());
//
//                Request request = new Request.Builder()
//                        .url(urlCheckIpBatch)
//                        .post(requestBody)
//                        .build();
//
//                Response response = client.newCall(request).execute();
//
//                if (response.isSuccessful()) {
//                    // Обработка успешного ответа
//                    final JSONArray jsonResponse = new JSONArray(response.body().string());
//                    h.post(() -> {
//                        if (setIpInfo1(jsonResponse, serverList)) {
//                            callback.onResponse(serverList);
//                        }
//                    });
//                } else {
//                    // Обработка ошибки
//                    final String errorMessage = response.message();
//                    h.post(() -> callback.onError(errorMessage));
//                }
//            } catch (Exception e) {
//                e.printStackTrace();
//                final String errorMessage = e.getMessage();// Обработка исключений
//                h.post(new Runnable() {
//                    @Override
//                    public void run() {
//                        callback.onError(errorMessage);
//                    }
//                });
//            }
//        });
//    }
//
//    private boolean setIpInfo1(JSONArray response, List<ProxyInfo> serverList) {
//        boolean result = false;
//
//        for (int i = 0; i < response.length(); i++) {
//            try {
//                // ip ==> ipInfo.get("query").toString()
//                JSONObject ipInfo = response.getJSONObject(i);
//                DLog.d("@@@@@@@" + ipInfo);
//
//                ProxyInfo server = serverList.get(i);
//                server.setCity(ipInfo.getString(Config.KEY_CITY));
//                server.setRegionName(ipInfo.getString(Config.KEY_REGION_NAME));
//                server.setLat(ipInfo.getDouble(Config.KEY_LAT));
//                server.setLon(ipInfo.getDouble(Config.KEY_LON));
//                server.code = ipInfo.getString("countryCode");
//
//                if (server.ip.length() < 6) {
//                    server.ip = ipInfo.getString("query");
//                }
//
//
//                server.country = ipInfo.getString("country");
//                server.region = ipInfo.getString("region");
//                server.zip = ipInfo.getString("zip");
//
//                server.timezone = ipInfo.getString("timezone");
//                server.isp = ipInfo.getString("isp");
//                server.org = ipInfo.getString("org");
//                server.as = ipInfo.getString("as");
//                result = true;
//            } catch (JSONException e) {
//                result = false;
//                DLog.handleException(e);
//            }
//        }
//        return result;
//    }
//
//
//    public boolean setIpInfo(JSONArray response, List<MtprotoProxy> serverList) {
//        boolean result = false;
//
//        for (int i = 0; i < response.length(); i++) {
//            try {
//                // ip ==> ipInfo.get("query").toString()
//                JSONObject ipInfo = response.getJSONObject(i);
//                DLog.d("@@@@@@@" + ipInfo);
//
//                MtprotoProxy server = serverList.get(i);
//                server.setCity(ipInfo.getString(Config.KEY_CITY));
//                server.setRegionName(ipInfo.getString(Config.KEY_REGION_NAME));
//                server.setLat(ipInfo.getDouble(Config.KEY_LAT));
//                server.setLon(ipInfo.getDouble(Config.KEY_LON));
//                server.code = ipInfo.getString("countryCode");
//
//
//                server.country = ipInfo.getString("country");
//                server.region = ipInfo.getString("region");
//                server.zip = ipInfo.getString("zip");
//
//                server.timezone = ipInfo.getString("timezone");
//                server.isp = ipInfo.getString("isp");
//                server.org = ipInfo.getString("org");
//                server.as = ipInfo.getString("as");
//                result = true;
//            } catch (JSONException e) {
//                result = false;
//                DLog.handleException(e);
//            }
//        }
//        return result;
//    }
//}

File: mtproto/src/main/java/com/walhalla/mtprotolist/manager/MUtils.kt
package com.walhalla.mtprotolist.manager

import android.os.Handler
import com.walhalla.mtprotolist.Config
import com.walhalla.mtprotolist.entity.MtprotoProxy
import com.walhalla.mtprotolist.utils.QueryFormat.getArray0
import com.walhalla.mtprotolist.utils.QueryFormat.getQueryArray
import com.walhalla.mtprotolist.webproxy.ProxyInfo
import com.walhalla.ui.DLog.d
import com.walhalla.ui.DLog.handleException
import okhttp3.MediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.util.concurrent.Executor
import java.util.concurrent.Executors

class MUtils(private val h: Handler) {
    private val executor: Executor = Executors.newCachedThreadPool()

    interface Callback<T> {
        fun onError(error: String?)

        fun onResponse(serverList: T?)
    }

    private val client = OkHttpClient()
    private val urlCheckIpBatch = "http://ip-api.com/batch"

    fun getIpInfo(
        serverList: List<MtprotoProxy>,
        callback: Callback<List<MtprotoProxy>>,
    ) {
        executor.execute {
            try {
                val jsonArray = getArray0(serverList)
                if (jsonArray.length() != serverList.size) {
                    h.post { callback.onError("Host resolve failed") }
                    return@execute
                }
                val JSON = MediaType.parse("application/json; charset=utf-8")
                val requestBody = RequestBody.create(JSON, jsonArray.toString())

                val request = Request.Builder()
                    .url(urlCheckIpBatch)
                    .post(requestBody)
                    .build()

                val response = client.newCall(request).execute()

                if (response.isSuccessful()) {
                    val jsonResponse = JSONArray(response.body()!!.string())
                    h.post {
                        val geoError = applyGeoResponseToMtproto(jsonResponse, serverList)
                        if (geoError == null) {
                            callback.onResponse(serverList)
                        } else {
                            callback.onError(geoError)
                        }
                    }
                } else {
                    val errorMessage = response.message()
                    h.post { callback.onError(errorMessage) }
                }
            } catch (e: Exception) {
                handleException(e)
                val errorMessage = e.message
                h.post { callback.onError(errorMessage) }
            }
        }
    }

    fun getIpInfoProxyInfo(
        serverList: MutableList<ProxyInfo>,
        callback: Callback<List<ProxyInfo>>,
    ) {
        executor.execute {
            try {
                val jsonArray = getQueryArray(serverList)
                if (jsonArray.length() != serverList.size) {
                    h.post { callback.onError("Host resolve failed") }
                    return@execute
                }
                val JSON = MediaType.parse("application/json; charset=utf-8")
                val requestBody = RequestBody.create(JSON, jsonArray.toString())

                val request = Request.Builder()
                    .url(urlCheckIpBatch)
                    .post(requestBody)
                    .build()

                val response = client.newCall(request).execute()

                if (response.isSuccessful()) {
                    val jsonResponse = JSONArray(response.body()!!.string())
                    h.post {
                        val geoError = applyGeoResponseToProxyInfo(jsonResponse, serverList)
                        if (geoError == null) {
                            callback.onResponse(serverList)
                        } else {
                            callback.onError(geoError)
                        }
                    }
                } else {
                    val errorMessage = response.message()
                    h.post { callback.onError(errorMessage) }
                }
            } catch (e: Exception) {
                handleException(e)
                val errorMessage = e.message
                h.post { callback.onError(errorMessage) }
            }
        }
    }

    private fun applyGeoResponseToProxyInfo(response: JSONArray, serverList: MutableList<ProxyInfo>): String? {
        if (response.length() != serverList.size) {
            return "Geo lookup failed"
        }

        for (i in 0..<response.length()) {
            val ipInfo = response.getJSONObject(i)
            d("@@@@@@@" + ipInfo)

            val geoError = validateGeoResponse(ipInfo)
            if (geoError != null) {
                return geoError
            }

            try {
                applyGeoToProxyInfo(serverList[i], ipInfo)
            } catch (e: JSONException) {
                handleException(e)
                return "Geo lookup failed"
            }
        }
        return null
    }

    private fun applyGeoResponseToMtproto(response: JSONArray, serverList: List<MtprotoProxy>): String? {
        if (response.length() != serverList.size) {
            return "Geo lookup failed"
        }

        for (i in 0..<response.length()) {
            val ipInfo = response.getJSONObject(i)
            d("@@@@@@@$ipInfo")

            val geoError = validateGeoResponse(ipInfo)
            if (geoError != null) {
                return geoError
            }

            try {
                applyGeoToMtproto(serverList[i], ipInfo)
            } catch (e: JSONException) {
                handleException(e)
                return "Geo lookup failed"
            }
        }
        return null
    }

    private fun validateGeoResponse(ipInfo: JSONObject): String? {
        when (ipInfo.optString("status")) {
            "fail" -> return ipInfo.optString("message", "Geo lookup failed")
            "success" -> Unit
            else -> if (ipInfo.optString("country").isBlank()) {
                return "Geo lookup failed"
            }
        }

        if (ipInfo.optString("country").isBlank() && ipInfo.optString("countryCode").isBlank()) {
            return "Empty geo data"
        }

        return null
    }

    private fun applyGeoToProxyInfo(server: ProxyInfo, ipInfo: JSONObject) {
        server.city = ipInfo.getString(Config.KEY_CITY)
        server.regionName = ipInfo.getString(Config.KEY_REGION_NAME)
        server.lat = ipInfo.getDouble(Config.KEY_LAT)
        server.lon = ipInfo.getDouble(Config.KEY_LON)
        server.code = ipInfo.getString("countryCode")

        if (server.ip!!.length < 6) {
            server.ip = ipInfo.getString("query")
        }

        server.country = ipInfo.getString("country")
        server.region = ipInfo.getString("region")
        server.zip = ipInfo.getString("zip")
        server.timezone = ipInfo.getString("timezone")
        server.isp = ipInfo.getString("isp")
        server.org = ipInfo.getString("org")
        server.`as` = ipInfo.getString("as")
    }

    private fun applyGeoToMtproto(server: MtprotoProxy, ipInfo: JSONObject) {
        server.city = ipInfo.getString(Config.KEY_CITY)
        server.regionName = ipInfo.getString(Config.KEY_REGION_NAME)
        server.lat = ipInfo.getDouble(Config.KEY_LAT)
        server.lon = ipInfo.getDouble(Config.KEY_LON)
        server.code = ipInfo.getString("countryCode")
        server.country = ipInfo.getString("country")
        server.region = ipInfo.getString("region")
        server.zip = ipInfo.getString("zip")
        server.timezone = ipInfo.getString("timezone")
        server.isp = ipInfo.getString("isp")
        server.org = ipInfo.getString("org")
        server.`as` = ipInfo.getString("as")
    }
}
